app/src/utils/index.tsx

```diff
@@ -0,0 +1,97 @@
+import React from 'react'
+import { toaster } from 'toasterhea'
+import BigNumber from 'bignumber.js'
+import InsufficientFundsError from '$shared/errors/InsufficientFundsError'
+import getNativeTokenName from '$shared/utils/nativeToken'
+import Toast, { ToastType } from '$shared/toasts/Toast'
+import { fromAtto } from '$mp/utils/math'
+import { getProjectRegistryContract } from '$app/src/getters'
+import { Layer } from '$utils/Layer'
+import getPublicWeb3 from '$utils/web3/getPublicWeb3'
+
+export async function ensureGasMonies(
+    chainId: number,
+    account: string,
+    { recover = false }: { recover?: boolean } = {},
+) {
+    while (true) {
+        try {
+            const balance = fromAtto(
+                new BigNumber(await getPublicWeb3(chainId).eth.getBalance(account)),
+            )
+
+            if (balance.isGreaterThan(0)) {
+                break
+            }
+
+            throw new InsufficientFundsError(account)
+        } catch (e) {
+            if (recover && e instanceof InsufficientFundsError) {
+                const tokenName = getNativeTokenName(chainId)
+
+                try {
+                    await toaster(Toast, Layer.Toast).pop({
+                        title: 'Not enough tokens',
+                        type: ToastType.Warning,
+                        desc: (
+                            <p>
+                                You need {tokenName} for gas. Get some
+                                <br />
+                                and try again.
+                            </p>
+                        ),
+                        okLabel: 'Try again',
+                        cancelLabel: 'Cancel',
+                    })
+
+                    continue
+                } catch (_) {
+                    throw e
+                }
+            }
+
+            throw e
+        }
+    }
+}
+
+/**
+ * Scouts for the `Subscribed` event associated with the Project Registry contract
+ * and explodes if it can't find one after given number of tries.
+ * @param chainId The chain id related to the purchase.
+ * @param projectId Project id.
+ * @param account Account who made the purchase.
+ * @param options.attempts Number of tries (with 3s delay between each).
+ * @returns Nothing important. Watch for 'splosions tho.
+ */
+export async function waitForPurchasePropagation(
```

Comment:
This is a rewrite of the old utility with a similar name (dropped in this PR). The new one is cleaner and is based on attempts rather than pure time-based timeouts.